File: frontend/src/lib/types/character.ts
/**
 * Character TypeScript type definitions
 * Matches backend schemas from /backend/src/shinkei/schemas/character.py
 */

export type EntityImportance = "major" | "minor" | "background";

export interface Character {
	id: string;
	world_id: string;
	name: string;
	description?: string;
	aliases?: string[];
	role?: string;
	importance: EntityImportance;
	first_appearance_beat_id?: string;
	custom_metadata?: Record<string, any>;
	created_at: string;
	updated_at: string;
}

export interface CharacterCreate {
	name: string;
	description?: string;
	aliases?: string[];
	role?: string;
	importance?: EntityImportance;
	first_appearance_beat_id?: string;
	custom_metadata?: Record<string, any>;
}

export interface CharacterUpdate {
	name?: string;
	description?: string;
	aliases?: string[];
	role?: string;
	importance?: EntityImportance;
	first_appearance_beat_id?: string;
	custom_metadata?: Record<string, any>;
}

export interface CharacterResponse extends Character {}

export interface CharacterListResponse {
	characters: CharacterResponse[];
	total: number;
	page: number;
	page_size: number;
}

export interface CharacterWithMentionsResponse extends CharacterResponse {
	mention_count: number;
}

export interface CharacterSearchResponse {
	characters: CharacterWithMentionsResponse[];
	total: number;
}

// UI Helper types
export const ENTITY_IMPORTANCE_LABELS: Record<EntityImportance, string> = {
	major: "Major Character",
	minor: "Minor Character",
	background: "Background Character"
};

export const ENTITY_IMPORTANCE_COLORS: Record<EntityImportance, string> = {
	major: "bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-300",
	minor: "bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300",
	background: "bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-300"
};

export const ENTITY_IMPORTANCE_OPTIONS: { value: EntityImportance; label: string }[] = [
	{ value: "major", label: "Major Character" },
	{ value: "minor", label: "Minor Character" },
	{ value: "background", label: "Background Character" }
];

// Common character roles for suggestions
export const COMMON_CHARACTER_ROLES = [
	"Protagonist",
	"Antagonist",
	"Mentor",
	"Ally",
	"Rival",
	"Love Interest",
	"Comic Relief",
	"Sidekick",
	"Guide",
	"Guardian",
	"Trickster",
	"Herald"
];

File: backend/src/shinkei/schemas/character.py
"""Character Pydantic schemas for API validation."""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict
from shinkei.security.validators import PlainText, SanitizedHTML


class CharacterBase(BaseModel):
    """Base character schema with common fields."""
    name: PlainText(200) = Field(..., min_length=1)
    description: Optional[SanitizedHTML(10000)] = None
    aliases: Optional[list[PlainText(100)]] = Field(None, max_length=10)
    role: Optional[PlainText(200)] = None
    importance: str = Field(default="background", pattern="^(major|minor|background)$")
    first_appearance_beat_id: Optional[str] = None
    custom_metadata: Optional[dict] = None


class CharacterCreate(CharacterBase):
    """Schema for creating a new character."""
    pass


class CharacterUpdate(BaseModel):
    """Schema for updating a character."""
    model_config = ConfigDict(extra='forbid')

    name: Optional[PlainText(200)] = Field(None, min_length=1)
    description: Optional[SanitizedHTML(10000)] = None
    aliases: Optional[list[PlainText(100)]] = Field(None, max_length=10)
    role: Optional[PlainText(200)] = None
    importance: Optional[str] = Field(None, pattern="^(major|minor|background)$")
    first_appearance_beat_id: Optional[str] = None
    custom_metadata: Optional[dict] = None


class CharacterResponse(CharacterBase):
    """Schema for character responses."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    world_id: str
    created_at: datetime
    updated_at: datetime


class CharacterListResponse(BaseModel):
    """Schema for paginated character list."""
    characters: list[CharacterResponse]
    total: int
    page: int
    page_size: int


class CharacterWithMentionsResponse(CharacterResponse):
    """Schema for character with mention count."""
    mention_count: int = 0


class CharacterSearchResponse(BaseModel):
    """Schema for character search results."""
    characters: list[CharacterWithMentionsResponse]
    total: int
